### ISSUE TITLE
Stuck on Setting the controller to learning mode...

### ISSUE THREAD

**Issue by BenJamesAndo** [EduardoHerreraM/SmartIR-Zigbee-Tools#2]
Hello. Thanks so much for making this!
I've got Python 3.11.7 running on Windows 11. I have all the dependencies installed.
When I run `poetry run python app/main.py` it then returns:
```Set the device to `off` mode```
```Setting the controller to learning mode...```

But nothing after that happens. My UFO-R11 doesn't go into learning mode. Am I doing something wrong? I've also tried it with Python 3.10 and Python 3.12. I'm also running Mosquitto and Zigbee2MQTT. I have `poetry 1.7.1` installed.
I do have moes_ufo_r11_controller.py in \app\src\domain\controllers and have set my env variables.

My best guess is that it's not connecting to Mosquitto at all. Nothing shows up in the logs of Zigbee2MQTT. If I run the following service in Home Assistant it does go into learning mode
```
service: mqtt.publish
data:
  topic: zigbee2mqtt/moes_ufo_r11/set
  payload: "{\"learn_ir_code\": \"ON\"}"
```

**Comment by EduardoHerreraM** [comment]
> Hello. Thanks so much for making this! I've got Python 3.11.7 running on Windows 11. I have all the dependencies installed. When I run `poetry run python app/main.py` it then returns: `` Set the device to `off` mode `` `Setting the controller to learning mode...`
> 
> But nothing after that happens. My UFO-R11 doesn't go into learning mode. Am I doing something wrong? I've also tried it with Python 3.10 and Python 3.12. I'm also running Mosquitto and Zigbee2MQTT. I have `poetry 1.7.1` installed. I do have moes_ufo_r11_controller.py in \app\src\domain\controllers and have set my env variables.
> 
> My best guess is that it's not connecting to Mosquitto at all. Nothing shows up in the logs of Zigbee2MQTT. If I run the following service in Home Assistant it does go into learning mode
> 
> ```
> service: mqtt.publish
> data:
>   topic: zigbee2mqtt/moes_ufo_r11/set
>   payload: "{\"learn_ir_code\": \"ON\"}"
> ```

Hello! Thanks for using this :).

Yep, it totally looks like the connection is not being stablished.  The project is missing some logs for this.

I'll try to update the project, checking the connection when the app is executed first. I'll let you know :).

**Comment by EduardoHerreraM** [comment]
@BenJamesAndo can you try the code from this branch?

` git checkout feat/check_mqtt_connection `

I've been doing some tests and most likely is an authentication problem. The library is not so well written in regarding of error handling, so it's hard to know for me right now.

Go inside your Zigbee2MQTT configuration path and extract the defined username and password that is there. The path for me is: 

`/config/z2mqtt/configuration.yaml`

Here, search for `mqtt`, and `user` and `password`.

Then fulfill the `USERNAME` and `PASSWORD` inside `SmartIR-Zigbee-Tools/properties.env` with those values.

Let me know if that works for you :).

**Comment by BenJamesAndo** [comment]
I gave the new branch a go. No error logs showed up. Now I'm thinking it might be that the dependencies aren't installed properly. Perhaps it thinks they are but they aren't and as such does nothing.
If I run `python paho-mqtt` then I get `[Errno 2] No such file or directory`

No matter what I've tried so far I can't get it to use my environment variables. I've been mostly using Windows 11 but I've also tried Debian with the same issue. Running in Windows 11 with Visual Studio Code works, but ends up with `Setting the controller to learning mode...`
So it's perhaps not an issue with this project but rather my setup.
Perhaps if all the dependencies were included in the folder that may help?

I've successfully connected to my MQTT from another external program so wrong credentials aren't an issue.

**Comment by BenJamesAndo** [comment]
Or am I supposed to instead of manually installing the dependencies I run with `poetry install` first? If I do that
```
Installing the current project: smartir-zigbee-tools (0.1.0)
The current project could not be installed: No file/folder found for package smartir-zigbee-tools
If you do not want to install the current project use --no-root
```

Sorry I'm not super good with python or poetry.

**Comment by EduardoHerreraM** [comment]
@BenJamesAndo I'll try to create a Makefile and include instructions of how to execute the project. 

You should be able to execute it with `poetry run python app/main.py`. But I'll try to create the Makefile.

**Comment by EduardoHerreraM** [comment]
@BenJamesAndo I've updated the branch.

Try running `make install` and `make run`.

If the connection is correct, then you will see the normal project output. If not, then an error will be raised.

Let me know if you have more problems :)

**Comment by BenJamesAndo** [comment]
Perhaps some progress on my end. I now get this when I run `make run`

`Traceback (most recent call last):
File "app/main.py", line 4, in <module>
from src.application.generate_climate_json import GenerateClimateJson
File "/home/ben/Downloads/SmartIR-Zigbee-Tools-feat-check_mqtt_connection/app/
src/application/generate_climate_json.py", line 7, in <module>
class GenerateClimateJson:
File "/home/ben/Downloads/SmartIR-Zigbee-Tools-feat-check_mqtt_connection/app/
src/application/generate_climate_json.py", Line 37, in GenerateClimateJson
operation_modes: list[str],
TypeError: 'type' object is not subscriptable
make: *** [Makefile:10: run] Error 1`

This is my .env file with ip address and password changed for me pasting it here.
```
BROKER_ADDRESS=192.168.1.xxx
BROKER_PORT=1883
USERNAME=mqtt-user
PASSWORD=password
CONTROLLER_FRIENDLY_NAME=moes_ufo_r11
MANUFACTURER=Samsung
MODEL=AR09TXHQASIN
MINIMUM_TEMPERATURE=18.0
MAXIMUM_TEMPERATURE=30.0
OPERATION_MODES=["auto", "cool", "dry", "heat"]
FAN_MODES=["auto", "low", "medium", "high"]
SWING_MODES=["on"]
```

**Comment by EduardoHerreraM** [comment]
@BenJamesAndo That totally looks like you are using a different python version.

Can you run `poetry run python -V` and check that you are using `3.11`??

You can fix your problem deleting the typing of the functions too.

**Comment by BenJamesAndo** [comment]
Hey. Sorry for my late response, I've been out of town.
Yes you're right. I was testing it lately on a different Linux VM and it was running Python 3.8.
I've just spun up a new Debian 12 install which comes with Python 3.11. This is what I get when I run `make install`
```Installing dependencies from lock file

Package operations: 0 installs, 1 update, 0 removals

  • Updating typing-extensions (4.9.0 -> 4.8.0): Failed

  Error

  g-io-error-quark: The connection is closed (18)

  at /usr/lib/python3/dist-packages/keyring/backends/libsecret.py:134 in get_credential
      130│         Otherwise, it will return the first username and password combo that it finds.
      131│         """
      132│         query = self._query(service, username)
      133│         try:
    → 134│             items = Secret.password_search_sync(
      135│                 self.schema, query, Secret.SearchFlags.UNLOCK, None
      136│             )
      137│         except GLib.Error as error:
      138│             quark = GLib.quark_try_string('g-io-error-quark')

make: *** [Makefile:4: install] Error 1
```

When I run `make run` I get
```
Traceback (most recent call last):
  File "/home/ben/Downloads/SmartIR-Zigbee-Tools-feat-check_mqtt_connection/app/main.py", line 16, in <module>
    mqtt_client = MosquittoMQTTClient(
                  ^^^^^^^^^^^^^^^^^^^^
  File "/home/ben/Downloads/SmartIR-Zigbee-Tools-feat-check_mqtt_connection/app/src/infrastructure/mosquitto_mqtt_client.py", line 30, in __init__
    self.__check_connection_established()
  File "/home/ben/Downloads/SmartIR-Zigbee-Tools-feat-check_mqtt_connection/app/src/infrastructure/mosquitto_mqtt_client.py", line 42, in __check_connection_established
    raise ConnectionNotEstablishedException()
src.domain.exceptions.ConnectionNotEstablishedException: Connection cannot be established. Most likely, the username and password are not correct.
make: *** [Makefile:10: run] Error 1
```

I definitely have the username and password correct. But it's good to see the error messages now showing up.

When I first ran `make install` it didn't work so I had to manually install the dependencies via commands such as `apt install python3-paho-mqtt`

Perhaps I need to learn more about python virtual environments and run it that way?
What I can also do it is try to get paho-mqtt running on it's own so I can troubleshoot any issues it may be facing.

**Comment by EduardoHerreraM** [comment]
@BenJamesAndo So, you are running a VM, hmm.. That shouldn't be an issue if it can connect to the server.

I can't diagnose the problem you are facing with poetry. Perhaps I shouldn't be using it, is an advanced tool for virtual environment, but maybe too advance for the purpose of the project.

Nonetheless, if you installed the dependencies with the packet manager, then it's ok. Not recommended, but ok.

So I can't really give you more information than the error that is being shown. You may have to create another user-password with `moquitto_passwd` or something. 

You aren't having connection problems to the server. If the initial connection can't be established  (as it's unavailable) then another error is raised, a `TimeoutError`. The error that is raised is my custom-made error, but `paho` doesn't give much information about what the error is..

TLDR; create a new user-password in mosquitto and set those in the properties. You most-likely are facing a credentials problem.

**Comment by BenJamesAndo** [comment]
You are right. It was a credentials issue when using the Debian VM. The username & password that's in `/config/z2mqtt/configuration.yaml` for me didn't work. I'm running Mosquitto broker and Zigbee2MQTT as Home Assistant addons. As such all Home Assistant users are compatible Mosquito broker users. I had tested one of the Home Assistant users in the past but it didn't work back then, perhaps due to a firewall issue.
Thank you so very much for your help.

For Windows 11 I can't get it to work. It says the credentials are incorrect even though I'm using the same .env file as on the Debian VM. I'll just stick with Debian for now. I had to run as root user and also install `dbus-x11`
